lib/src/Uncertainty/Algorithm/WeightedExperiments/SmolyakExperiment.cxx

```diff
@@ -0,0 +1,348 @@
+    }
+  } // Loop over (potentially) duplicated nodes, weights
+  // Extract the map
+  const UnsignedInteger size = nodeWeightMap.size();
+  LOGDEBUG(OSS() << "size = " <<  size);
+  LOGDEBUG(OSS() << "Extract the map");
+  nodes = Sample(size, dimension);
+  weights = Point(size);
+  if (size > 0)
+  {
+    UnsignedInteger index = 0;
+    for (std::map<Point, Scalar>::iterator it = nodeWeightMap.begin(); it != nodeWeightMap.end(); ++ it)
+    {
+      LOGDEBUG(OSS() << "[" << index << "], add " << it->first << " = " << it->second);
+      nodes[index] = it->first;
+      weights[index] = it->second;
+      ++ index;
+    } // Loop over unique nodes, weights in map
+  }
+}
+
+/* Compute the nodes and weights */
+/*         This may involve negative weights.
+
+The algorithm has 3 steps:
+- create the multi-index set for the combination technique,
+- create the list of elementary Smolyak quadrature corresponding
+  to each multi-index in the set,
+- merge the elementary quadratures to avoid duplicate nodes,
+  updating the weights if necessary.
+
+The algorithm to merge the elementary quadratures
+starts with an empty sample of unique nodes Q^U and weights w^U.
+For each candidate node, we search if it is already in the
+sample of unique nodes:
+- if the node is not found in Q^U, it is added to Q^U,
+  and the weight is added to w^U,
+- otherwise, the weight is updated.
+*/
+Sample SmolyakExperiment::generateWithWeights(Point & weights) const
+{
+  LOGDEBUG(OSS() << "SmolyakExperiment::computeNodesAndWeights()");
+  const UnsignedInteger dimension = collection_.getSize();
+  LOGDEBUG(OSS() << "  dimension = " << dimension);
+  if (!(dimension > 0))
+    throw InvalidArgumentException(HERE) << "Error: expected a positive number of marginal experiments, here it is " << dimension;
+  // Compute multiindex set
+  const IndicesCollection combinationIndicesCollection(computeCombination());
+  LOGDEBUG(OSS() << "  combinationIndicesCollection = " << combinationIndicesCollection);
+  // Create elementary Smolyak quadratures
+  Sample nodes(0, dimension);
+  weights.clear();
+  const UnsignedInteger numberOfUnitaryQuadratures = combinationIndicesCollection.getSize();
+  for (UnsignedInteger i = 0; i < numberOfUnitaryQuadratures; ++ i)
+  {
+    WeightedExperimentCollection collection(collection_);
+    for (UnsignedInteger j = 0; j < dimension; ++ j)
+      collection[j].setSize(combinationIndicesCollection(i, j));
+    TensorProductExperiment elementaryExperiment(collection);
+    Point elementaryWeights(0);
+    const Sample elementaryNodes(elementaryExperiment.generateWithWeights(elementaryWeights));
+    // Compute Smolyak coefficient
+    UnsignedInteger marginalLevelsSum = 0;
+    for (UnsignedInteger j = 0; j < dimension; ++ j)
+      marginalLevelsSum += combinationIndicesCollection(i, j);
+    const UnsignedInteger exponent = level_ + dimension - marginalLevelsSum - 1;
+    const Scalar smolyakSign = (exponent % 2 == 0) ? 1.0 : -1.0;
```

Comment:
Something like 1.0 - 2.0 * (exponent % 2) to avoid a test and a branching point.

Reply:
this seems clever, but adds two operations and is less readable, so I keep my version